openmdao/solvers/linear/tests/test_direct_solver.py

```diff
@@ -101,6 +102,114 @@ def test_rev_mode_bug(self):
 
+    def test_raise_error_on_singular(self):
+        prob = Problem()
+        prob.model = model = Group()
+
+        comp = IndepVarComp()
+        comp.add_output('dXdt:TAS', val=1.0)
+        comp.add_output('accel_target', val=2.0)
+        model.add_subsystem('des_vars', comp, promotes=['*'])
+
+        teg = model.add_subsystem('thrust_equilibrium_group', subsys=Group())
+        teg.add_subsystem('dynamics', ExecComp('z = 2.0*thrust'), promotes=['*'])
+
+        thrust_bal = BalanceComp()
+        thrust_bal.add_balance(name='thrust', val=1207.1, lhs_name='dXdt:TAS',
+                               rhs_name='accel_target', eq_units='m/s**2', lower=-10.0, upper=10000.0)
+
+        teg.add_subsystem(name='thrust_bal', subsys=thrust_bal,
+                          promotes_inputs=['dXdt:TAS', 'accel_target'],
+                          promotes_outputs=['thrust'])
+
+        teg.linear_solver = DirectSolver()
+
+        teg.nonlinear_solver = NewtonSolver()
+        teg.nonlinear_solver.options['solve_subsystems'] = True
+        teg.nonlinear_solver.options['max_sub_solves'] = 1
+        teg.nonlinear_solver.options['atol'] = 1e-4
+
+        prob.setup(check=False)
+        prob.set_solver_print(level=0)
+
+        with self.assertRaises(RuntimeError) as cm:
+            prob.run_model()
+
+        expected_msg = "Singular entry found in 'thrust_equilibrium_group' for state/residual 'thrust'."
+
+        self.assertEqual(expected_msg, str(cm.exception))
+
+    def test_raise_error_on_singular_with_densejac(self):
+        prob = Problem()
+        prob.model = model = Group()
+
+        comp = IndepVarComp()
+        comp.add_output('dXdt:TAS', val=1.0)
+        comp.add_output('accel_target', val=2.0)
+        model.add_subsystem('des_vars', comp, promotes=['*'])
+
+        teg = model.add_subsystem('thrust_equilibrium_group', subsys=Group())
+        teg.add_subsystem('dynamics', ExecComp('z = 2.0*thrust'), promotes=['*'])
+
+        thrust_bal = BalanceComp()
+        thrust_bal.add_balance(name='thrust', val=1207.1, lhs_name='dXdt:TAS',
+                               rhs_name='accel_target', eq_units='m/s**2', lower=-10.0, upper=10000.0)
+
+        teg.add_subsystem(name='thrust_bal', subsys=thrust_bal,
+                          promotes_inputs=['dXdt:TAS', 'accel_target'],
+                          promotes_outputs=['thrust'])
+
+        teg.linear_solver = DirectSolver()
+        teg.jacobian = DenseJacobian()
+
+        teg.nonlinear_solver = NewtonSolver()
+        teg.nonlinear_solver.options['solve_subsystems'] = True
+        teg.nonlinear_solver.options['max_sub_solves'] = 1
+        teg.nonlinear_solver.options['atol'] = 1e-4
+
+        prob.setup(check=False)
+        prob.set_solver_print(level=0)
+
+        with self.assertRaises(RuntimeError) as cm:
+            prob.run_model()
+
+        expected_msg = "Singular entry found in 'thrust_equilibrium_group' for state/residual 'thrust'."
```

Comment:
is it possible to be more precise with this error? Can we tell them whether its a row or a column thats singular?

Reply:
Well, numpy doesn't give the info, but we could return the norm of the row and column so that the user can decide what to do.